### ISSUE TITLE
[BUG] Cannot login to Dynamics due issue in Login method

### ISSUE THREAD

**Issue by Shakevg** [Microsoft/EasyRepro#752]
## Bug Report
Login method fails due trying to click to not existing controls, but do not check that is is exists before

**EasyRepro Version**
  - [ ] Microsoft Dynamics 365 Online Version 8.2 (8.2.x) (DB 8.2.x) online
  - [ ] Microsoft Dynamics 365 Online Version 9.0 (9.0.x) (DB 9.0.x) online
  - [ ] Microsoft Dynamics 365 Online Version 9.0.2 (9.0.2) (DB 9.0.2) online
  - [ + ] Microsoft Dynamics 365 Online Version 9.1 (9.1.x) (DB 9.1.x) online
  
**UCI or Classic Web**
  - [ + ] UCI (XrmApp)
  - [ ] Classic Web (XrmBrowser)

**Online or On Premise**
  - [ + ] Online
  - [ ] On Premise

**Browser**
  - [ + ] Chrome
  - [ ] Firefox
  - [ ] IE
  - [ ] Edge

**Describe the bug**  
During login test trying to click to not existing controls, but do not check that is is exists before

```csharp
driver.ClickWhenAvailable(By.Id("aadTile"));
driver.ClickWhenAvailable(By.Id("use_another_account_link"));
```
Stack Trace:
```
Test method Microsoft.Dynamics365.UIAutomation.Sample.UCI.OpenContactUCI.UCITestOpenContactRetrieveHeaderValues threw exception: 
System.InvalidOperationException: Unable to click element.
    at Microsoft.Dynamics365.UIAutomation.Browser.SeleniumExtensions.<>c__DisplayClass50_0.<WaitUntilClickable>b__0() in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\Extensions\SeleniumExtensions.cs:line 500
   at Microsoft.Dynamics365.UIAutomation.Browser.SeleniumExtensions.WaitUntil(ISearchContext driver, Func`2 searchFunc, Nullable`1 timeout, Action`1 successCallback, Action failureCallback) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\Extensions\SeleniumExtensions.cs:line 558
   at Microsoft.Dynamics365.UIAutomation.Browser.SeleniumExtensions.WaitUntilClickable(ISearchContext driver, By by, Nullable`1 timeout, Action`1 successCallback, Action failureCallback) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\Extensions\SeleniumExtensions.cs:line 508
   at Microsoft.Dynamics365.UIAutomation.Browser.SeleniumExtensions.WaitUntilClickable(ISearchContext driver, By by, Nullable`1 timeout, Action`1 successCallback, String exceptionMessage) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\Extensions\SeleniumExtensions.cs:line 500
   at Microsoft.Dynamics365.UIAutomation.Browser.SeleniumExtensions.ClickWhenAvailable(ISearchContext driver, By by, Nullable`1 timeout, String errorMessage) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\Extensions\SeleniumExtensions.cs:line 38
   at Microsoft.Dynamics365.UIAutomation.Api.UCI.WebClient.Login(IWebDriver driver, Uri uri, SecureString username, SecureString password, SecureString mfaSecrectKey, Action`1 redirectAction) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Api.UCI\WebClient.cs:line 127
   at Microsoft.Dynamics365.UIAutomation.Browser.DelegateBrowserCommand`6.ExecuteCommand(IWebDriver driver, Object[] params) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\DelegateBrowserCommand.cs:line 163
   at Microsoft.Dynamics365.UIAutomation.Browser.BrowserCommand`1.Execute[T1,T2,T3,T4,T5,T6,T7,T8,T9](IWebDriver driver, T1 p1, T2 p2, T3 p3, T4 p4, T5 p5, T6 p6, T7 p7, T8 p8, T9 p9) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\BrowserCommand.cs:line 136
   at Microsoft.Dynamics365.UIAutomation.Browser.BrowserCommand`1.Execute[T1,T2,T3,T4,T5](IWebDriver driver, T1 p1, T2 p2, T3 p3, T4 p4, T5 p5) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\BrowserCommand.cs:line 66
   at Microsoft.Dynamics365.UIAutomation.Browser.BrowserPage.Execute[TResult,T1,T2,T3,T4,T5](BrowserCommandOptions options, Func`7 delegate, T1 p1, T2 p2, T3 p3, T4 p4, T5 p5) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Browser\BrowserPage.cs:line 242
   at Microsoft.Dynamics365.UIAutomation.Api.UCI.WebClient.Login(Uri orgUri, SecureString username, SecureString password, SecureString mfaSecrectKey, Action`1 redirectAction) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Api.UCI\WebClient.cs:line 116
   at Microsoft.Dynamics365.UIAutomation.Api.UCI.OnlineLogin.Login(Uri orgUrl, SecureString username, SecureString password, SecureString mfaSecrectKey) in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Api.UCI\Elements\OnlineLogin.cs:line 38
   at Microsoft.Dynamics365.UIAutomation.Sample.UCI.OpenContactUCI.UCITestOpenContactRetrieveHeaderValues() in C:\BCD\EasyRepro-develop.747\Microsoft.Dynamics365.UIAutomation.Sample\UCI\Read\OpenContact.cs:line 150
```

**Code to reproduce**  
Run UCITestOpenContactRetrieveHeaderValues test

```csharp
xrmApp.OnlineLogin.Login(_xrmUri, _username, _password);
```
 
**Expected behavior**  
It should be verification that controls exists before click
```csharp
if (driver.IsVisible(By.Id("aadTile")))
	driver.ClickWhenAvailable(By.Id("aadTile"));
if (driver.IsVisible(By.Id("use_another_account_link")))
	driver.ClickWhenAvailable(By.Id("use_another_account_link"
```));

**Comment by AngelRodriguez8008** [comment]
Fixed in #751 & also merged in #745

Thanks.

**Comment by TYLEROL** [comment]
@Shakevg - Per Angel's fixes that have been merged into develop, could you validate that your scenario is resolved by testing with the 'develop' branch?

Thanks,
Tyler

**Comment by Shakevg** [comment]
@TYLEROL, @AngelRodriguez8008 
Login is performed now but after Login method is hang on EnterOneTimeCode(driver, mfaSecrectKey); command inside Login method. 
Looks like it should be something like: 

```csharp
private LoginResult Login(IWebDriver driver, Uri uri, SecureString username, SecureString password, SecureString mfaSecrectKey = null, Action<LoginRedirectEventArgs> redirectAction = null)
        {
            bool online = !(OnlineDomains != null && !OnlineDomains.Any(d => uri.Host.EndsWith(d)));
            driver.Navigate().GoToUrl(uri);
....
            if(mfaSecrectKey != null)
            {
                EnterOneTimeCode(driver, mfaSecrectKey);
            }
.....
}
```

**Comment by AngelRodriguez8008** [comment]
Thanks @Shakevg, 

I'm getting some problems by fixing this.
In your way the Chrome Windows, for some estrange reason is keeping open.
![image](https://user-images.githubusercontent.com/27274469/74133355-b175ba80-4be8-11ea-861e-6cccf9ef176a.png)

I't taking some other considerations Like, waiting time to fail & error message.
I will keep you updated.

Thanks.

**Comment by AngelRodriguez8008** [comment]
Hi @Shakevg,

Thanks again for the Feedback.  I'm trying to simulate the situation but I not't able to tests without MFA. I will try to find one Env to test. Howerver, I think that my lastest [changes](https://github.com/microsoft/EasyRepro/pull/754/commits/128d8353c2d2c65edf493886decfe3db5106447a) should fix this problem.

I will keep it in PR #754, where I fix other Bug. @TYLEROL  let me know please if you need split this in two different PR.

Thanks & br

**Comment by AngelRodriguez8008** [comment]
Ok I glad to inform you that I have already tested in an Env. without MFA. It works perfectly. Running also pretty fast. :-)
![image](https://user-images.githubusercontent.com/27274469/74155267-8e610000-4c14-11ea-8ff4-48cf2f3e2e48.png)
the Fail is not because Login.

D365 Online 
2019 release wave 2 enabled
Server version: 9.1.0000.12935
Client version: 1.3.379-2001.2

**Comment by rsObjektkultur** [comment]
Can we get the fixes into the Nuget Package asap please?

My whole test suited is blocked by this failed login issue. And I can not simply roleback because the previous (working) version is no longer on Nuget available (only some very old versions, which have other breaking changes).

**Comment by AngelRodriguez8008** [comment]
Sorry @rsObjektkultur, I will test better the next time. Maybe can you compile the DLL by yourself in the middle time, until this get released.
